src/pm-conversations/pmConversationsModel.js

```diff
@@ -0,0 +1,158 @@
+// User IDs, excluding self, sorted numerically, joined with commas.
+export opaque type PmConversationKey = string;
+
+/** PRIVATE.  Exported only for tests. */
+// Input must have the exact right (multi-)set of users.  Needn't be sorted.
+export function keyOfExactUsers(ids: number[]): PmConversationKey {
+  return ids.sort((a, b) => a - b).join(',');
+}
+
+// Input may contain self or not, and needn't be sorted.
+function keyOfUsers(ids: number[], ownUserId: number): PmConversationKey {
+  return keyOfExactUsers(ids.filter(id => id !== ownUserId));
+}
+
+// Input must indeed be a PM, else throws.
+function keyOfPrivateMessage(msg: Message | Outbox, ownUserId: number): PmConversationKey {
+  return keyOfUsers(recipientsOfPrivateMessage(msg).map(r => r.id), ownUserId);
+}
+
+/** The users in the conversation, other than self. */
+export function usersOfKey(key: PmConversationKey): number[] {
+  return key ? key.split(',').map(s => Number.parseInt(s, 10)) : [];
+}
+
+//
+//
+// State and reducer.
+//
+
+/**
+ * The list of recent PM conversations, plus data to efficiently maintain it.
+ *
+ * This gets initialized from the `recent_private_conversations` data
+ * structure in the `/register` response (aka our initial fetch), and then
+ * kept up to date as we learn about new or newly-fetched messages.
+ */
+// (Compare the webapp's implementation, in static/js/pm_conversations.js.)
+export type PmConversationsState = {|
+  // The latest message ID in each conversation.
+  map: Immutable.Map<PmConversationKey, number>,
+
+  // The keys of the map, sorted by latest message descending.
+  sorted: Immutable.List<PmConversationKey>,
+|};
+
+const initialState: PmConversationsState = { map: Immutable.Map(), sorted: Immutable.List() };
+
+// Insert the key at the proper place in the sorted list.
+//
+// Optimized, taking O(1) time, for the case where that place is the start,
+// because that's the common case for a new message.  May take O(n) time in
+// general.
+function insertSorted(sorted, map, key, msgId) {
+  const i = sorted.findIndex(k => {
+    const id = map.get(k);
+    invariant(id !== undefined, 'pm-conversations: key in sorted should be in map');
+    return id < msgId;
+  });
+
+  if (i === 0) {
+    return sorted.unshift(key);
+  }
+  if (i < 0) {
+    return sorted.push(key);
+  }
+  return sorted.insert(i, key);
+}
+
+// Can take linear time in general.  That sounds inefficient...
+// but it's what the webapp does, so must not be catastrophic. 🤷
+function insert(
+  state: PmConversationsState,
+  key: PmConversationKey,
+  msgId: number,
+): PmConversationsState {
+  let { map, sorted } = state;
+  const prev = map.get(key);
+  if (prev === undefined) {
+    // The conversation is new.  Add it.
+    map = map.set(key, msgId);
+    return { map, sorted: insertSorted(sorted, map, key, msgId) };
+    //
+  } else if (prev >= msgId) {
+    // The conversation already has a newer message.  Do nothing.
+    return state;
+    //
```

Comment:
I wanted a blank line here so that the three cases would be visually separated; but Prettier wants to push them back together again. This was a way of making space... though not a great way, as the `//` really diminishes the visual separation.

I'll try just `prettier-ignore`ing this section.